knime_extension/src/nodes/locationanalysis.py

```diff
@@ -633,3 +681,548 @@ def getIndex(s):
+        port_index=1,
+        column_filter=knut.is_int_or_string,
+        include_row_key=False,
+        include_none_column=False,
+    )
+    threshold = knext.DoubleParameter(
+        "Threshold",
+        "Desired threshold distance or cost",
+    )
+    pnumber = knext.IntParameter(
+        "Optimum p number ",
+        "The optimum p number of facilities",
+        4,
+    )
+
+    def configure(self, configure_context, input_schema_1, input_schema_2):
+        return knext.Schema.from_columns(
+            [
+                knext.Column(knext.string(), _FACILITYID),
+                knext.Column(knext.int32(), _Chosen),
+            ]
+        )
+
+    def execute(self, exec_context: knext.ExecutionContext, input_1, input_2):
+        import pulp
+        import pandas as pd
+
+        df1, df2 = locationadata.load(input_1, self.id_left, input_2, self.id_right)
+
+        popu = df1[self.populationid].values.tolist()
+        cutoff = self.threshold
+
+        cost_matrix, demand, candidate, p, nrequire = locationadata.process(
+            df1, df2, self.pnumber, self.requiredid
+        )
+
+        # Define the optimization problem
+        prob = pulp.LpProblem("mclp", pulp.LpMinimize)
+
+        # Define the decision variables
+        x, y = locationadata.ppulpxy(demand, candidate)
+
+        # Define the objective function
+        prob += pulp.lpSum(
+            [x[(i, j)] * popu[i] for i in range(demand) for j in range(candidate)]
+        )
+
+        # Define the constraints
+        for i in range(demand):
+            prob += pulp.lpSum([x[(i, j)] for j in range(candidate)]) == 1
+
+        for j in range(candidate):
+            for i in range(demand):
+                prob += x[(i, j)] <= y[j]
+
+        # Define the threshold cutoff distance constraint
+        for i in range(demand):
+            prob += (
+                pulp.lpSum(
+                    [cost_matrix.iloc[i][j] * x[(i, j)] for j in range(candidate)]
+                )
+                <= cutoff
+            )
+
+        # Define the one facility required constraint
+        prob += pulp.lpSum([y[j] for j in range(candidate)]) == p
+        prob += y[nrequire] == 1
+        # Solve the problem
+        prob.solve()
+
+        status = pulp.LpStatus[prob.status]
+        if status != "Optimal":
+            raise RuntimeError(f"Solver failed with status: {status}")
+
+        # Extract the value of y for each candidate facility
+        dfy = pd.DataFrame({_FACILITYID: cost_matrix.columns})
+        for j in range(candidate):
+            dfy.loc[j, _Chosen] = pulp.value(y[j])
+        dfy[_FACILITYID] = dfy[_FACILITYID].astype(str)
+        dfy[_Chosen] = dfy[_Chosen].astype(int)
+        return knext.Table.from_pandas(dfy)
+
+
+############################################
+# Location-allocation LSCP Solver
+############################################
+@knext.node(
+    name="LSCP Solver",
+    node_type=knext.NodeType.MANIPULATOR,
```

Comment:
NodeType.PREDICTOR like the other solver nodes?

Reply:
it is solve some equation to get result, I am not sure